## Exploration notebook
This notebook is used in order to explore and get a a deeper understanding of the columns in the file `event_datafile_new.csv`.
Having a fine comprehension of each feature will help to select the columns that should be:
1. The partition key;
2. The clustering columns (if needed);
3. The data columns that will be also added to the partition.

In [1]:
import pandas as pd

In [2]:
# read the datafile 
df = pd.read_csv("event_datafile_new.csv")

In [3]:
df.head()

artist firstName gender  \
0  Barry Tuckwell/Academy of St Martin-in-the-Fie...  Mohammad      M   
1                                       Jimi Hendrix  Mohammad      M   
2                                       Building 429  Mohammad      M   
3                                         The B-52's    Gianna      F   
4                                   Die Mooskirchner    Gianna      F   

   itemInSession   lastName     length level  \
0              0  Rodriguez  277.15873  paid   
1              1  Rodriguez  239.82975  paid   
2              2  Rodriguez  300.61669  paid   
3              0      Jones  321.54077  free   
4              1      Jones  169.29914  free   

                                  location  sessionId  \
0  Sacramento--Roseville--Arden-Arcade, CA        961   
1  Sacramento--Roseville--Arden-Arcade, CA        961   
2  Sacramento--Roseville--Arden-Arcade, CA        961   
3    New York-Newark-Jersey City, NY-NJ-PA        107   
4    New York-Newark-Jersey City, NY-NJ-PA        107   

                                                song  userId  
0  Horn Concerto No. 4 in E flat K495: II. Romanc...      88  
1                            Woodstock Inprovisation      88  
2                               Majesty (LP Version)      88  
3                                         Love Shack      38  
4                                  Frisch und g'sund      38

In [4]:
df.shape

(6820, 11)

### Exploring the dataset to create a table able to perform queries to answer the _first question._
"1. Give me the artist, song title and song's length in the music app history that was heard during  sessionId = 338, and itemInSession = 4"

In [5]:
# Unique values in sessionId column
df.sessionId.nunique()

776

In [7]:
# Unique values in itemInSession column
df.itemInSession.nunique()

123

In [13]:
type(df.sessionId.values[0])

numpy.int64

In [14]:
# creating a new column to check the nunique combinations generated by the combination of the two columns
df["sessionId_&_itemInSession"] = df.sessionId.astype(str) + df.itemInSession.astype(str)

In [17]:
df["sessionId_&_itemInSession"].nunique()

6806

As there are 6820 rows and 6806 unique columns in the combination, there will be needed to have at leats three columns in the primary key for in order to perform the query for the first question properly.   
It could be done using any other column that must be included in the table (artist, song title, song length). **But, to assure a robust data model, song length will be used as the second clustering column as it is less likely to have another song with the exctly same length in the same sessionId and itemInSession.**

### Exploring the dataset to create a table able to perform queries to answer the _second question._
"2. Give me only the following: name of artist, song (sorted by itemInSession) and user (first and last name) for userid = 10, sessionid = 182"